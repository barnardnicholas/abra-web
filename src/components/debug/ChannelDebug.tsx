import React, { useState } from 'react';
import { LastPlayed, SoundsTally, SoundsTallyItem } from '../../types/Debug';
import { SoundChannel, soundTypes } from '../../types/Scenario';
import FileDebug from './FileDebug';

interface ChannelProps {
  channel: SoundChannel;
}

function ChannelDebug({ channel }: ChannelProps) {
  const { name, slug, isPlaying, durations, type, paths } = channel;

  const [lastPlayed, setLastPlayed] = useState<LastPlayed>(
    Object.keys(isPlaying).reduce((acc: LastPlayed, curr: string) => {
      return { ...acc, [curr]: null };
    }, {}),
  );

  const soundsTally: SoundsTally = paths.reduce((acc: SoundsTally, curr: string, i: number) => {
    const fileName: string = curr.split('/').pop() || '-';
    const fileSlug: string = fileName.split('.')[0];
    const newItem: SoundsTallyItem = {
      fileName,
      fullPath: curr,
      isPlaying: isPlaying[curr] || false,
      displayDuration: !Number.isNaN(durations[i]) ? `${(durations[i] / 1000).toFixed(1)}s` : '0',
    };
    return { ...acc, [fileSlug]: newItem };
  }, {} as SoundsTally);

  return (
    <div className="debug-item">
      <strong>{`${soundTypes[type]} Channel: ${name} (${slug})`}</strong>
      <div>
        <table className="debug-table">
          <thead>
            <tr>
              <td>FileName</td>
              <td>Is Playing</td>
              <td>Duration</td>
            </tr>
          </thead>
          <tbody>
            {Object.values(soundsTally).map((item: SoundsTallyItem, i: number) => {
              return (
                <FileDebug
                  key={item.fileName}
                  item={item}
                  lastPlayed={lastPlayed[item.fullPath]}
                  duration={durations[i]}
                  setLastPlayed={setLastPlayed}
                />
              );
            })}
          </tbody>
        </table>
      </div>
      <br />
    </div>
  );
}

export default ChannelDebug;
