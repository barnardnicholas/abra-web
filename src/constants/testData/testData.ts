import { Vector3 } from 'three';
import { SoundChannel, soundTypes, soundTypeValues } from '../../types/Scenario';

export const paths = ['slug1.mp3', 'slug2.mp3', 'slug3.mp3'];

export const testChannel: SoundChannel = {
  id: 0,
  name: 'Name',
  slug: 'slug',
  position: new Vector3(0, 0, 0),
  isPlaying: paths.reduce((acc: Record<string, boolean>, curr: string, index: number) => {
    return { ...acc, [curr]: index === 1 ? true : false };
  }, {}),
  durations: new Array(paths.length).fill(0),
  type: soundTypes.random,
  paths: paths.map((path: string) => `/audio/${soundTypeValues[soundTypes.random]}/${path}`),
  currentPath: paths[Math.floor(Math.random() * paths.length)] || paths[0],
  frequency: 0.5,
  area: [
    [0, 0, 0],
    [0, 0, 0],
  ],
  volume: 0.5,
  mute: false,
};
