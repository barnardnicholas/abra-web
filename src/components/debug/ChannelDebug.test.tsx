import React from 'react';
import renderer from 'react-test-renderer';
import { testChannel } from '../../constants/testData/testData';
import ChannelDebug from './ChannelDebug';

it('Renders correctly', () => {
  const tree = renderer.create(<ChannelDebug channel={testChannel} />).toJSON();
  expect(tree).toMatchSnapshot();
});
