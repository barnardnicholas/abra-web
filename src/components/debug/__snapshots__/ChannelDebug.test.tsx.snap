// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div
  className="debug-item"
>
  <strong>
    random Channel: Name (slug)
  </strong>
  <div>
    <table
      className="debug-table"
    >
      <thead>
        <tr>
          <td>
            FileName
          </td>
          <td>
            Is Playing
          </td>
          <td>
            Duration
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            slug1.mp3
          </td>
          <td>
            <div
              className="play-indicator "
              style={
                Object {
                  "transition": "max-width 0ms linear",
                }
              }
            />
          </td>
          <td>
            0.0s
          </td>
        </tr>
        <tr>
          <td>
            slug2.mp3
          </td>
          <td>
            <div
              className="play-indicator "
              style={
                Object {
                  "transition": "max-width 0ms linear",
                }
              }
            />
          </td>
          <td>
            0.0s
          </td>
        </tr>
        <tr>
          <td>
            slug3.mp3
          </td>
          <td>
            <div
              className="play-indicator "
              style={
                Object {
                  "transition": "max-width 0ms linear",
                }
              }
            />
          </td>
          <td>
            0.0s
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <br />
</div>
`;
